# Regression Optimizations

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# sklearn
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score, root_mean_squared_error

from sklearn.preprocessing import PolynomialFeatures

In [2]:
# Load the dataset as a pandas DataFrame
california = fetch_california_housing(as_frame=True)
X = california.data
y = california.target

X.head()
y.head()

0    4.526
1    3.585
2    3.521
3    3.413
4    3.422
Name: MedHouseVal, dtype: float64

## Train-test split

In [3]:
# from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=432)

## Baseline Metrics
Using default hyper-parameters for LinearRegression()

In [4]:
model = LinearRegression()
model.fit(X_train,y_train) 

y_pred = model.predict(X_test)

# Diagnose your model
mae = mean_absolute_error(y_test, y_pred)   # MAE is the average absolute error
mse = mean_squared_error(y_test, y_pred)    # MSE is mean squared error
rmse = np.sqrt(mse)                         # is it larger because of the unit scale?
r2 = r2_score(y_test, y_pred)               # R² shows how much variance the model explains - model captures 59.6% of the pattern

print(f"MAE:  {mae*100000:,.0f}")
#print(f"MSE:  {mse:.3f}")
print(f"RMSE: {rmse*100000:,.0f}")
print(f"R²:   {r2:.8f}")

MAE:  52,669
RMSE: 71,529
R²:   0.60802296


In [5]:
y.describe()['50%']*100000, y.describe()['mean']*100000

(np.float64(179700.0), np.float64(206855.81690891474))

# RESTART NOTEBOOK AND CLEAR ALL OUTPUTS

MAE:  52,669
RMSE: 71,529
R²:   0.60802296

In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# sklearn
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score, root_mean_squared_error
from sklearn.ensemble import RandomForestRegressor

from sklearn.preprocessing import PolynomialFeatures

In [7]:
# Load the dataset as a pandas DataFrame
california = fetch_california_housing(as_frame=True)
X = california.data
y = california.target

## Polynomial Linear Regression Metrics

To capture curved relationships that linear regression misses, and to check if accuracy improves.

When your data is nonlinear, residuals show patterns, or you want to compare model performance.

You use MAE, MSE, RMSE, and R² to see whether the polynomial model improves prediction accuracy versus the simple linear model.

In [8]:
print(X.shape)
poly = PolynomialFeatures(degree=2, include_bias=False)
X_poly = poly.fit_transform(X)
print(X_poly.shape)

(20640, 8)
(20640, 44)


In [9]:
# Explore the outputs

#df_poly = pd.DataFrame(X_poly, columns=poly.get_feature_names_out(X.columns))
#df_poly.head(7)
#df_poly.head().T

In [10]:
# from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=432)

In [11]:
%%time

model = RandomForestRegressor()
model.fit(X_train,y_train) 

y_pred = model.predict(X_test)

# Diagnose your model
mae = mean_absolute_error(y_test, y_pred)   # MAE is the average absolute error
mse = mean_squared_error(y_test, y_pred)    # MSE is mean squared error
rmse = np.sqrt(mse)                         # is it larger because of the unit scale?
r2 = r2_score(y_test, y_pred)               # R² shows how much variance the model explains - model captures 59.6% of the pattern

print(f"MAE:  {mae*100000:,.0f}")
#print(f"MSE:  {mse:.3f}")
print(f"RMSE: {rmse*100000:,.0f}")
print(f"R²:   {r2:.8f}")

MAE:  32,157
RMSE: 49,303
R²:   0.81377642
CPU times: total: 11.3 s
Wall time: 11.3 s
